whir/src/whir/batch/committer.rs

```diff
@@ -1,184 +1,248 @@
         &self,
-        merlin: &mut Merlin,
-        polys: &[CoefficientList<F::BasePrimeField>],
-    ) -> ProofResult<Witnesses<F, MerkleConfig>>
-    where
-        Merlin: FieldWriter<F> + FieldChallenges<F> + ByteWriter + DigestWriter<MerkleConfig>,
-    {
-        let timer = start_timer!(|| "Batch Commit");
-        let base_domain = self.0.starting_domain.base_domain.unwrap();
-        let expansion = base_domain.size() / polys[0].num_coeffs();
-        let expand_timer = start_timer!(|| "Batch Expand");
+        polys: witness::RowMajorMatrix<E::BaseField>,
+    ) -> Result<(Witnesses<E>, WhirCommitmentInTranscript<E>), Error> {
+        let mut transcript = BasicTranscript::<E>::new(b"commitment");
+        let polys = polys.to_mles();
+        let timer = entered_span!("Batch Commit");
+        let expansion = self.0.starting_domain.size() / polys[0].evaluations().len();
+        let expand_timer = entered_span!("Batch Expand");
         let evals = polys
             .par_iter()
-            .map(|poly| expand_from_coeff(poly.coeffs(), expansion))
+            .map(|poly| {
+                expand_from_coeff(
+                    &match poly.evaluations() {
+                        #[cfg(feature = "parallel")]
+                        FieldType::Base(evals) => {
+                            let mut evals = evals
+                                .par_iter()
+                                .map(|e| E::from_base(e))
+                                .collect::<Vec<_>>();
+                            interpolate_over_boolean_hypercube(&mut evals);
+                            evals
+                        }
+                        #[cfg(not(feature = "parallel"))]
+                        FieldType::Base(evals) => {
+                            let mut evals =
+                                evals.iter().map(|e| E::from_base(e)).collect::<Vec<_>>();
+                            interpolate_over_boolean_hypercube(&mut evals);
+                            evals
+                        }
+                        FieldType::Ext(evals) => {
+                            let mut evals = evals.clone();
+                            interpolate_over_boolean_hypercube(&mut evals);
+                            evals
+                        }
+                        _ => panic!("Invalid field type"),
+                    },
+                    expansion,
+                )
+            })
             .collect::<Vec<Vec<_>>>();
-        end_timer!(expand_timer);
+        exit_span!(expand_timer);
 
-        assert_eq!(base_domain.size(), evals[0].len());
+        assert_eq!(self.0.starting_domain.size(), evals[0].len());
 
         // These stacking operations are bottleneck of the commitment process.
         // Try to finish the tasks with as few allocations as possible.
-        let stack_evaluations_timer = start_timer!(|| "Stack Evaluations");
+        let stack_evaluations_timer = entered_span!("Stack Evaluations");
```

Comment:
Does it possible we combine 
- stack_evaluations
- restructure_evaluations
- horizontal_stacking
in one shot? it might relax performance a bit

Reply:
Yes, it is possible.